### PR TITLE
add extract api

### PR DESCRIPTION
**Pull request by dholth**
Is Enum a good idea?

### CODE REVIEW

File: conda_package_streaming/exceptions.py [hunk 1]
```diff
@@ -0,0 +1,21 @@
+class InvalidArchiveError(Exception):
+    """Raised on OSError"""
+
+    def __init__(self, msg, *args, **kw):
+        msg = (
+            "Error with archive. You probably need to delete and re-download "
+            "or re-create this file. Message was:\n\n%s" % (msg,)
+        )
+        super(InvalidArchiveError, self).__init__(msg)
```
Comment (jezdez) [inline conda_package_streaming/exceptions.py:9]:
```suggestion
        super().__init__(msg)
```

File: conda_package_streaming/exceptions.py [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+class CaseInsensitiveFileSystemError(Exception):
+    def __init__(self):
+        message = """\
+Cannot extract package to a case-insensitive file system. Your install
+destination does not differentiate between upper and lowercase characters, and
+this breaks things. Try installing to a location that is case-sensitive. Windows
+drives are usually the culprit here - can you install to a native Unix drive, or
+turn on case sensitivity for this (Windows) location?
+        """
+        super(CaseInsensitiveFileSystemError, self).__init__(message)
```
Comment (jezdez) [inline conda_package_streaming/exceptions.py:21]:
```suggestion
        super().__init__(message)
```

**Review by jezdez — APPROVED** [review]
LGTM minus nits.

**Comment by jezdez** [comment]
> Is Enum a good idea?

🤷

File: conda_package_streaming/exceptions.py (hunk 1 above)
Comment (dholth) [reply conda_package_streaming/exceptions.py]: pre-commit seems more "aggressive" on futurize than I remember. Not a fan. But it does automatically make this change.

File: conda_package_streaming/extract.py [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+            else:
+                # tarfile ExtractError instead / OSError subclass?
+                raise exceptions.InvalidArchiveError(
+                    f"failed with error: {str(e)}"
+                ) from e
+
+        # next iteraton of for loop raises GeneratorExit in stream
+        stream.close()
+
+    if sys.platform.startswith("linux") and os.getuid() == 0:
```
Comment (dholth) [inline conda_package_streaming/extract.py:64]: this would be terrible if dest_dir was / but conda will never call it that way

**Comment by dholth** [comment]
I will probably merge this next week. I dislike is the exceptions module, which is copied from `conda-package-handling`. Maybe they can inherit from `TarError` (for tarslip tarballs) and `OSError` (for case sensitivity errors).

**Pull request merged**